### PR TITLE
[FLINK-30239][rest] Fix the problem that FlameGraph cannot be generated due to using ImmutableSet incorrectly

### PR DESCRIPTION
**Pull request by 1996fanrui**
## What is the purpose of the change

Fix the problem that FlameGraph cannot be generated due to using ImmutableSet incorrectly

## Brief change log

- The first commit : Change the log level to error when flame graph generation fails and refactor some codes (It's useful to trouble-shooting int the future)
- The second commit: Remove the ImmutableSet to fix the bug and improve the unit test

## Verifying this change

This change added tests and can be verified as follows:

  - *JobVertexThreadInfoTrackerTest#testGroupExecutionsByLocation*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not documented

### CODE REVIEW

**Comment by 1996fanrui** [comment]
Hi @afedulov  @fapaul , this is a bug about flame graph and it's caused by #19228 [FLINK-26074].

Please help take a look in your free time, thanks~

**Comment by 1996fanrui** [comment]
Hi @xintongsong  , this is a bug about flame graph and it's caused by #19228 [FLINK-26074].

The flame graph doesn't work due this bug after flink 1.16, please help take a look in your free time, thanks~

File: flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/JobVertexThreadInfoTracker.java [hunk 1]
```diff
@@ -251,7 +251,7 @@ private void triggerThreadInfoSampleInternal(
         return executionsWithGateways;
     }
 
-    private Map<TaskManagerLocation, ImmutableSet<ExecutionAttemptID>> groupExecutionsByLocation(
+    Map<TaskManagerLocation, ImmutableSet<ExecutionAttemptID>> groupExecutionsByLocation(
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/JobVertexThreadInfoTracker.java:254]: Having to change the method visibility for testing is usually an indicator of testing against internal implementation. We should treat the testee as blackbox and test against its contract. In this case, the contract should be that `getVertexStats` should work properly when there's multiple executions for the same execution vertex.

Comment (xintongsong) [reply flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/JobVertexThreadInfoTracker.java:254]: And in cases where there's no good way for testing without accessing internal methods, we should add an annotation `@VisibleForTesting`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/JobVertexThreadInfoTracker.java [hunk 2]
```diff
@@ -203,7 +203,14 @@ private void triggerThreadInfoSampleInternal(
...
 
-            sample.whenCompleteAsync(new ThreadInfoSampleCompletionCallback(key, vertex), executor);
+            sample.whenCompleteAsync(new ThreadInfoSampleCompletionCallback(key, vertex), executor)
+                    .exceptionally(
+                            throwable -> {
+                                LOG.warn(
+                                        "An exception occurred while generating the flame graph.",
+                                        throwable);
+                                return null;
+                            });
```
[the remark this answers is not in the document]
Comment (1996fanrui) [reply flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/JobVertexThreadInfoTracker.java]: Hi @xintongsong , thanks for your review.

Sorry, I missed that log  before due to that log level is debug.

I changed the log level to error when flame graph generation fails, it's useful to trouble-shooting in the future.

image

**Review by xintongsong — APPROVED** [review]
Thanks @1996fanrui. LGTM. Merging this.

**Comment by 1996fanrui** [comment]
@xintongsong , thanks for your review!

**Pull request closed without merging**